### ISSUE TITLE
False positive for Pydantic Field(default_factory=...) with list[Literal[...]]

### ISSUE THREAD

**Issue by andreasoprani** [astral-sh/ty#3641]
### Summary

I encountered an invalid-assignment error while checking my codebase with ty. The same code is accepted by pyright and works at runtime with Pydantic.
The issue happens when a Pydantic model has a field typed as a list of literals and uses default_factory to provide a default list containing one of those literal values.
Interestingly, the equivalent dataclass example does not produce an error.
Additionally, ty accepts the Pydantic version when the default factory is a separately defined function with an explicit return annotation.

Reproducible example:
```
from typing import Literal
from dataclasses import dataclass, field

from pydantic import BaseModel, Field

TestLit = Literal["a", "b", "c"]


def factory() -> list[TestLit]:
    return ["b"]


@dataclass
class TestClass:
    lit: list[TestLit] = field(default_factory=lambda: ["a"])


class TestPydantic(BaseModel):
    lit_a: list[TestLit] = Field(default_factory=lambda: ["a"])
    lit_b: list[TestLit] = Field(default_factory=factory)


print(TestClass().lit)  # prints ["a"]
print(TestPydantic().lit_a)  # prints ["a"]
print(TestPydantic().lit_b)  # prints ["b"]
```

Ty output:
```
error[[invalid-assignment](https://ty.dev/rules#invalid-assignment)]: Object of type `dataclasses.Field[list[str]]` is not assignable to `list[Literal["a", "b", "c"]]`
  --> test.py:19:12
   |
19 |     lit_a: list[TestLit] = Field(default_factory=lambda: ["a"])
   |            -------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Incompatible value of type `dataclasses.Field[list[str]]`
   |            |
   |            Declared type
   |

Found 1 diagnostic
```

Additional info:
```
❯ uv pip list
Package           Version
----------------- -------
annotated-types   0.7.0
pydantic          2.13.4
pydantic-core     2.46.4
typing-extensions 4.15.0
typing-inspection 0.4.2

❯ uvx ty --version
ty 0.0.42 (7f8cb6457 2026-06-01)

❯ uv run pyright test.py
0 errors, 0 warnings, 0 informations

❯ uv run pyright --version
pyright 1.1.408
```

### Version

ty 0.0.42 (7f8cb6457 2026-06-01)

**Comment by carljm** [comment]
Thanks for the report! At first glance this looks like something that should just be fixed in our dataclasses-transform implementation, it shouldn't require dedicated pydantic support. Not immediately sure why we are able to thread through the type context in the one case but not in the other.

**Comment by charliermarsh** [comment]
Empirically, it appears that https://github.com/astral-sh/ruff/pull/24832 fixes this?
